File: alanEappen/assignments/fastSlow/lc141/lc141.js
// Problem Statement #

//Given the head of a Singly LinkedList, write a function to determine if the LinkedList has a cycle in it or not.
//Given:
//head
//Goal: to return a boolean to see if the linkedlist has a cycle.
//variables:
//slowPtr, fastPtr is equal to the head
//fastPtr and fastPtr.next exists proceed with the loop
//slowPtr.next() 
//fastPtr.next().next()
//if(slowPtr == fastPtr)
//return true
//if not a cycle 
//return false
class Node {
  constructor(value, next = null) {
    this.value = value;
    this.next = next;
  }
}

const has_cycle = function (head) {
  let slowPtr = head;
  let fastPtr = head;

  while (fastPtr && fastPtr.next) {

    slowPtr = slowPtr.next
    fastPtr = fastPtr.next.next
    if (slowPtr === fastPtr) {
      return true
    }
  }

  return false
}


head = new Node(1)
head.next = new Node(2)
head.next.next = new Node(3)
head.next.next.next = new Node(4)
head.next.next.next.next = new Node(5)
head.next.next.next.next.next = new Node(6)
console.log(`LinkedList has cycle: ${has_cycle(head)}`)


head.next.next.next.next.next.next = head.next.next
console.log(`LinkedList has cycle: ${has_cycle(head)}`)

head.next.next.next.next.next.next = head.next.next.next
console.log(`LinkedList has cycle: ${has_cycle(head)}`)




// Solution
// -----
// function has_cycle(head) {
//   let slow = head,
//     fast = head;
//   while (fast !== null && fast.next !== null) {
//     fast = fast.next.next;
//     slow = slow.next;
//     if (slow === fast) {
//       return true; // found the cycle
//     }
//   }
//   return false;
// }
// -----

// Time Complexity #
// As we have concluded above, once the slow pointer enters the cycle, the fast pointer will meet the slow pointer in the same loop. Therefore, the time complexity of our algorithm will be O(N) where ‘N’ is the total number of nodes in the LinkedList.

// Space Complexity #
// The algorithm runs in constant space O(1).

File: wilsonChan/assignments/fastSlow/lc141/lc141.py
# Problem Statement #

# Given the head of a Singly LinkedList, write a function to determine if the LinkedList has a cycle in it or not.

'''

***Works on leetcode***

NameError: name 'has_cycle' is not defined



Variables:
-slow pointer (hop once)
-fast pointer (hop twice)


Steps:
-Initialize both slow pointer and fast pointer to head
-Use while loop to increment slow by one and increment fast by two
-When slow pointer = fast pointer, a loop has been made

'''



class Node:
  def __init__(self, value, next=None):
    self.value = value
    self.next = next


  def has_cycle(head):
  # TODO: Write your code here

    slow, fast = head, head

    while(fast and fast.next):
      slow = slow.next
      fast = fast.next.next
      
      if(slow == fast):
        return True

    return False


def main():
  head = Node(1)
  head.next = Node(2)
  head.next.next = Node(3)
  head.next.next.next = Node(4)
  head.next.next.next.next = Node(5)
  head.next.next.next.next.next = Node(6)
  print("LinkedList has cycle: " + str(has_cycle(head)))

  head.next.next.next.next.next.next = head.next.next
  print("LinkedList has cycle: " + str(has_cycle(head)))

  head.next.next.next.next.next.next = head.next.next.next
  print("LinkedList has cycle: " + str(has_cycle(head)))


main()

# Solution
# -----
# def has_cycle(head):
#   slow, fast = head, head
#   while fast is not None and fast.next is not None:
#     fast = fast.next.next
#     slow = slow.next
#     if slow == fast:
#       return True  # found the cycle
#   return False

# -----

# Time Complexity #
# As we have concluded above, once the slow pointer enters the cycle, the fast pointer will meet the slow pointer in the same loop. Therefore, the time complexity of our algorithm will be O(N) where ‘N’ is the total number of nodes in the LinkedList.

# Space Complexity #
# The algorithm runs in constant space O(1).
